0003-I-dapr-compose.md

```diff
@@ -0,0 +1,294 @@
+
+This increases the complexity of using dapr in development, where users want to run multiple services in local mode and be able to partially/fully replicate the production sceanrio.
+
+In K8s mode this is alleviated through the use of helm/deployment YAML files. There is currently no such capability available for local self hosted mode.
+
+Asking a user to run multiple different `dapr run` commands each with different flags, increases the complxity for users onboarding onto Dapr.
+
+### Why dapr CLI?
+
+From the initial [proposal](https://github.com/dapr/community/issues/207), the solution was proposed as a seprate repo and CLI in itself. But later it was suggested to use dapr CLI itself to have a `compose` command in it instead of a having a separate CLI.
+The main reason for including it in the dapr CLI itself is that, users do not have to download and use a new CLI in addition to the dapr CLI.
+
+This feature is tightly coupled and opinionated on how `dapr` is going to be run locally, and having a separate CLI `dapr-compose` deciding on how `dapr` CLI should be used, is not a good pattern to start with.
+
+> Note: `daprd` is more generally used and considered a binary and not necessarily a CLI tool. So `dapr` CLI is not making use of another CLI but rather passing on configs for running a binary.
+
+
+## Terminology
+
+- `dapr compose` refers to the whole feature.
+- `compose` refers to the CLI command that will added or the `compose` configuration file (will be specified on each use).
+- `compose` configuration file refers to the file required to use this feature whereas the `config_file` that is defined is the `Configuration` CRD that `daprd` uses.
+
+## Related Items
+
+### Related Proposals
+- https://github.com/dapr/community/issues/207
+- https://github.com/dapr/cli/issues/1123
+
+### Related Issues
+
+## Expectations and alternatives
+
+The scope of this proposal is to define a `compose` CLI command which will allow users to define and run multiple services from a single `compose` configuration file.
+
+This proposal specifically targets running in local environments and `slim` mode, where container engines are available. For running `daprd` container along with `app` container the solution is to use Kubernetes or docker-compose.
+
+For this proposal we will targetting runnning the applications and side cars as processes in the OS.
+
+Pros/Cons of using processes:
+- Pro: All other commands in `dapr` CLI for self hosted mode are written to work with processes
+- Con: Process execution is OS specific and it does not provide any isolation guarantees
+
+
+## Requirements
+
+The main requirements for the command:
+- being able to configure multiple dapr apps from a single configuration file
+- once the command is run with designated configuration file, if any application crashes, it will be restarted with an exponential backoff
```

Comment:
> if any application crashes, it will be restarted with an exponential backoff

I'm worried we will be creating and maintaining a process orchestrator here when the main goal of this is to provide a better developer experience to get started.

Reply:
Agreed. This proposal should play no part in restarting a failed application. It should be up to the developer to see what apps are running and be able to issue a run command to simply restart any failed apps. Otherwise, you are building a process monitor/orchestrator.

Reply:
I added this part, since it was mentioned as one of the main requirements in the initial community proposal itself. https://github.com/dapr/community/issues/207. First point mentions keeping track of termination and restart policies. 

I completely agree that if we go this route we will be going for a process orchestrator and the idea behind the _daemon monitor_ process mentioned as one of the options is that only. 

@artursouza thoughts?